File: LeetCode75_Artur/Array and String/1071_Greatest_Common_Divisor_of_Strings.cpp
#include <iostream>

using namespace std;

class Solution
{
public:
    size_t gcd(size_t a, size_t b)
    {
        if (a < b)
            return gcd(b, a);

        if (! (a % b))
            return b;
        else
            return gcd(b, a % b);
    }

    string gcdOfStrings(string str1, string str2)
    {
        size_t maxLen = gcd(str1.size(), str2.size());
        size_t subStrLen = 0;

        while(subStrLen < maxLen && str1[subStrLen] == str2[subStrLen])
            subStrLen++;

        if(!subStrLen)
            return "";

        size_t i = subStrLen, j = subStrLen;
        while(i < str1.size() || j < str2.size())
        {
            if(i < str1.size() && str1[i] != str1[i % subStrLen])
                return "";

            if(j < str2.size() && str2[j] != str2[j % subStrLen])
                return "";

            i++;
            j++;
        }

        return str1.substr(0, subStrLen);
    }
};

File: LeetCode75_Artur/Array and String/1071_Greatest_Common_Divisor_of_Strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        str1_len, str2_len = len(str1), len(str2)
        if str1_len >= str2_len:
            gcd = Solution.gcd(str1_len, str2_len)
            if str2[:gcd] * int(str1_len / gcd) == str1:
                return str2[:gcd]
        return ""

    def gcd(num1, num2):
        if num1 < num2:
            return Solution.gcd(num2, num1)
        elif num1 % num2 == 0:
            return num2
        return Solution.gcd(num2, num1 % num2)

Solution.gcdOfStrings()